### PR TITLE
Add Store Plugins

### PR DESCRIPTION
**Pull request by abettadapur**
* TODO: Add unit tests for the plugin system

### CODE REVIEW

File: packages/redux-dynamic-modules-saga/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import { default as createSagaMiddleware, SagaMiddleware } from "redux-saga";
+import { IPlugin, IItemManager, getRefCountedManager, IModuleManager } from "redux-dynamic-modules";
+import { ISagaRegistration, ISagaModule } from "./Contracts";
+import { getSagaManager } from "./SagaManager";
+import { sagaEquals } from "./SagaComparer";
+
+export function getSagaPlugin<C>(sagaContext?: C): IPlugin {
+    // setup the saga middleware
+    const sagaMonitor = window["__SAGA_MONITOR_EXTENSION__"] || undefined;
```
Comment (navneetg-msft) [inline packages/redux-dynamic-modules-saga/src/index.ts]: Should we only add this in debug mode?

File: packages/redux-dynamic-modules-saga/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        },
+
+        onModuleRemoved: (module: ISagaModule<any>): void => {
+            if (module.sagas) {
+                _sagaManager.remove(module.sagas);
+            }
+        },
+
+        dispose: () => {
+            sagaMiddleware = null;
```
Comment (navneetg-msft) [inline packages/redux-dynamic-modules-saga/src/index.ts]: We should dispose sagaManager here which in turn should dispose all running sagas.

Comment (abettadapur) [reply packages/redux-dynamic-modules-saga/src/index.ts]: Yes, thanks

File: packages/redux-dynamic-modules/src/Contracts.ts [hunk 3]
```diff
@@ -35,13 +31,14 @@ export interface IModule<State> {
     finalActions?: AnyAction[];
 }
 
-export interface ISagaWithArguments<T> {
-    saga: (argument?: T) => Iterator<any>;
-    argument?: T;
+export interface IPlugin {
```
Comment (navneetg-msft) [inline packages/redux-dynamic-modules/src/Contracts.ts]: Extension?

File: packages/redux-dynamic-modules/src/ModuleStore.ts [hunk 4]
```diff
@@ -1,36 +1,41 @@
...
+export function configureStore<SagaContext, S1, S2, S3, S4, S5, S6, S7>(initialState: DeepPartial<S1 & S2 & S3 & S4 & S5 & S6 & S7>, plugins: IPlugin[], m1: IModule<S1>, m2: IModule<S2>, m3: IModule<S3>, m4: IModule<S4>, m5: IModule<S5>, m6: IModule<S6>, m7: IModule<S7>): IModuleStore<S1 & S2 & S3 & S4 & S5 & S6 & S7>;
+export function configureStore<SagaContext, S1, S2, S3, S4, S5, S6, S7, S8>(initialState: DeepPartial<S1 & S2 & S3 & S4 & S5 & S6 & S7 & S8>, plugins: IPlugin[], m1: IModule<S1>, m2: IModule<S2>, m3: IModule<S3>, m4: IModule<S4>, m5: IModule<S5>, m6: IModule<S6>, m7: IModule<S7>, m8: IModule<S8>): IModuleStore<S1 & S2 & S3 & S4 & S5 & S6 & S7 & S8>;
+export function configureStore<SagaContext, State>(initialState: DeepPartial<State>, plugins: IPlugin[], ...initialModules: IModule<any>[]): IModuleStore<State>;
+export function configureStore<SagaContext, State>(initialState: DeepPartial<State>, plugins: IPlugin[], ...initialModules: IModule<any>[]): IModuleStore<State> {
+  if (!plugins) {
+    plugins = [];
+  }
+
+  const pluginMiddleware = plugins.reduce(
```
Comment (navneetg-msft) [inline packages/redux-dynamic-modules/src/ModuleStore.ts]: pluginMiddlewares?

Comment (navneetg-msft) [reply packages/redux-dynamic-modules/src/ModuleStore.ts]: Should we also dedupe if any?

File: packages/redux-dynamic-modules/src/Managers/ModuleManager.ts [hunk 5]
```diff
@@ -1,23 +1,15 @@
...
 export interface IModuleManager<State> extends IItemManager<IModule<State>> {
     setDispatch: (dispatch: Dispatch<AnyAction>) => void;
     getReducer: (state: State, action: AnyAction) => State;
 }
 
-export function getModuleManager<SagaContext, State>(
-    sagaMiddleware: SagaMiddleware<SagaContext>,
-    middlewareManager: IItemManager<Middleware>): IModuleManager<State> {
-
+export function getModuleManager<State>(middlewareManager: IItemManager<Middleware>, plugins: IPlugin[]): IModuleManager<State> {
```
Comment (navneetg-msft) [inline packages/redux-dynamic-modules/src/Managers/ModuleManager.ts]: we should use default params here instead so callers need not pass empty array.

File: packages/redux-dynamic-modules-saga/src/index.ts (hunk 1 above)
Comment (abettadapur) [reply packages/redux-dynamic-modules-saga/src/index.ts]: Fixed

File: packages/redux-dynamic-modules/src/Managers/ModuleManager.ts (hunk 5 above)
Comment (abettadapur) [reply packages/redux-dynamic-modules/src/Managers/ModuleManager.ts]: This is not exposed

File: packages/redux-dynamic-modules/src/ModuleStore.ts (hunk 4 above)
Comment (abettadapur) [reply packages/redux-dynamic-modules/src/ModuleStore.ts]: Renamed to extensions, will rename the middlewares

File: packages/redux-dynamic-modules/src/ModuleStore.ts [hunk 6]
```diff
@@ -41,30 +46,41 @@ export function configureStore<SagaContext, State>(initialState: DeepPartial<Sta
...
+        addModule,
+        addModules
+      });
+    }
+  });
+
   store.dispose = () => {
     // get all added modules and remove them
     const allModules = modules.getItems();
     modules.remove(allModules);
```
[the remark this answers is not in the document]
Comment (abettadapur) [reply packages/redux-dynamic-modules/src/ModuleStore.ts]: done

File: packages/redux-dynamic-modules/src/Contracts.ts (hunk 3 above)
Comment (abettadapur) [reply packages/redux-dynamic-modules/src/Contracts.ts]: done

**Pull request merged**